src/webgpu/api/validation/render_pipeline/inter_stage.spec.ts

```diff
@@ -1,10 +1,337 @@
 export const description = `
-TODO: interface matching between vertex and fragment shader validation for createRenderPipeline:
-    - superset, subset, etc.
+Interface matching between vertex and fragment shader validation for createRenderPipeline:
+
+[1] TODO(dawn:1448) replace with t.device.limits.maxInterStageShaderVariables
 `;
 
 import { makeTestGroup } from '../../../../common/framework/test_group.js';
+import { assert, range } from '../../../../common/util/util.js';
 
 import { CreateRenderPipelineValidationTest } from './common.js';
 
-export const g = makeTestGroup(CreateRenderPipelineValidationTest);
+class InterStageMatchingValidationTest extends CreateRenderPipelineValidationTest {
+  getVertexStateWithOutputs(outputs: string[]): GPUVertexState {
+    return {
+      module: this.device.createShaderModule({
+        code: `
+        struct A {
+            ${outputs.join(',\n')},
+            @builtin(position) pos: vec4<f32>,
+        }
+        @vertex fn main() -> A {
+            var vertexOut: A;
+            vertexOut.pos = vec4<f32>(0.0, 0.0, 0.0, 1.0);
+            return vertexOut;
+        }
+        `,
+      }),
+      entryPoint: 'main',
+    };
+  }
+
+  getFragmentStateWithInputs(
+    inputs: string[],
+    hasBuiltinPosition: boolean = false
+  ): GPUFragmentState {
+    return {
+      targets: [{ format: 'rgba8unorm' }],
+      module: this.device.createShaderModule({
+        code: `
+        struct B {
+            ${inputs.join(',\n')},
+            ${hasBuiltinPosition ? '@builtin(position) pos: vec4<f32>' : ''}
+        }
+        @fragment fn main(fragmentIn: B) -> @location(0) vec4<f32> {
+            return vec4<f32>(1.0, 1.0, 1.0, 1.0);
+        }
+        `,
+      }),
+      entryPoint: 'main',
+    };
+  }
```

Comment:
These helpers are good. optional idea: I noticed from reading the tests that including the variable names could make parameterization a bit messier. You could omit those, e.g. find/replace a placeholder symbol like
`['@location(0) __: f32', '@location(1) __: f32']`
and generate the variable names inside these helpers.
```
let varName = 0;
const getVarName = () => `v${varName++}`;
...
inputs.map(s => s.replace('__', getVarName)
```